packages/cli/generators/repository/index.js

```diff
@@ -0,0 +1,294 @@
+        name: 'dataSourceClassName',
+        message: PROMPT_MESSAGE_DATA_SOURCE,
+        choices: datasourcesList,
+        when: this.artifactInfo.dataSourceClassName === undefined,
+        default: datasourcesList[0],
+        validate: utils.validateClassName,
+      },
+    ])
+      .then(props => {
+        debug(`props: ${inspect(props)}`);
+        Object.assign(this.artifactInfo, props);
+        return props;
+      })
+      .catch(err => {
+        debug(`Error during prompt for datasource name: ${err}`);
+        return this.exit(err);
+      });
+  }
+
+  async inferRepositoryType() {
+    let result = await this.isConnectorType(
+      KEY_VALUE_CONNECTOR,
+      this.artifactInfo.dataSourceClassName,
+    );
+
+    if (result) {
+      this.artifactInfo.repositoryTypeClass = KEY_VALUE_REPOSITORY;
+      this.artifactInfo.defaultTemplate = REPOSITORY_KV_TEMPLATE;
+    } else {
+      this.artifactInfo.repositoryTypeClass = DEFAULT_CRUD_REPOSITORY;
+      this.artifactInfo.defaultTemplate = REPOSITORY_CRUD_TEMPLATE;
+    }
+
+    // assign the data source name to the information artifact
+    let dataSourceName = this.artifactInfo.dataSourceClassName
+      .replace('Datasource', '')
+      .toLowerCase();
+
+    Object.assign(this.artifactInfo, {
+      dataSourceName: dataSourceName,
+    });
+  }
+
+  async promptModels() {
+    let modelList;
+    try {
+      modelList = await utils.getArtifactList(
+        this.artifactInfo.modelDir,
+        'model',
+      );
+    } catch (err) {
+      return this.exit(err);
+    }
+
+    if (_.isEmpty(modelList)) {
+      return this.exit(
+        `${ERROR_NO_MODELS_FOUND} ${this.artifactInfo.modelDir}.
+        ${chalk.yellow(
+          'Please visit http://loopback.io/doc/en/lb4/Repository-generator.html for information on how models are discovered',
+        )}`,
+      );
+    }
+
+    return this.prompt([
+      {
+        type: 'checkbox',
+        name: 'modelNameList',
+        message: PROMPT_MESSAGE_MODEL,
+        choices: modelList,
+        when: this.artifactInfo.modelNameList === undefined,
+        default: modelList[0],
+      },
+    ])
+      .then(props => {
+        debug(`props: ${inspect(props)}`);
+        Object.assign(this.artifactInfo, props);
+        return props;
+      })
+      .catch(err => {
+        debug(`Error during prompt for repository variables: ${err}`);
+        return this.exit(err);
+      });
+  }
+
+  async promptModelId() {
+    if (_.isEmpty(this.artifactInfo.modelNameList)) {
+      return this.exit(new Error(`${ERROR_NO_MODEL_SELECTED}`));
+    } else {
+      // iterate thru each selected model, infer or ask for the ID type
+      for (let item of this.artifactInfo.modelNameList) {
+        debug(`iterating thru ${item}`);
+        this.artifactInfo.modelName = item;
+        const prompts = [
+          {
+            type: 'list',
+            name: 'idType',
+            message: `What is the ID type of ${item} `,
```

Comment:
Can you use `typeof ${model}.prototype.id`? See e.g. TodoRepository: https://github.com/strongloop/loopback-next/blob/ac011f80b229b5dd5c69851ab14b39705b852629/examples/todo/src/repositories/todo.repository.ts#L10-L13

In the future, we may want to ask the user for the _name_ of the id property to use.

Regardless of that, I think we should not be duplicating the id type name in the repository class if possible.

What do you think?

Reply:
yes, I will use  `typeof ${model}.prototype.id` if the model has the id field and if it is set to the primary key.  To do that, I am trying to see the best way to do it, probably just read the file and inspect it or use ATS to analyze it more in depth and even get the id name from it.

thoughts?